### PR TITLE
Changed default allocation & filtering strategy

### LINKED ISSUE #2477: Target Allocator is not capable of scaling out with least-weighted strategy
### Component(s)

target allocator

## Description

I'm not sure whether this is a bug or an intentional implemention, however the `least-weighted` strategy of the Target allocator is not capable of scaling out in terms of re-distributing the existing targets in case of new collector addition.

## Steps to Reproduce

1. Configure 5 scrape targets and deploy 2 collectors.
2. Let the collectors get their targets assigned.
  - `collector1` -> `target1`, `target3`, `target5`
  - `collector2` -> `target2`, `target4`
3. Deploy a 3rd collector.
4. Check target distribution.

## Expected Result

Expected outcome would be re-distribution of the targets across all 3 instances.

- `collector1` -> `target1`, `target4`
- `collector2` -> `target2`, `target5`
- `collector3` -> `target3`

## Actual Result

No re-distribution is triggered and the 3rd collector does not get assigned any target.

- `collector1` -> `target1`, `target3`, `target5`
- `collector2` -> `target2`, `target4`
- `collector3` -> ` `

### Kubernetes Version

1.27.1

### Operator version

0.90.0

### Collector version

0.91.0

### Log output

_No response_

### Implementation analysis

I've checked the code and if I'm not mistaken scaling out is not possible with the current implementation (please correct me if I'm wrong).

When newly added collector instances are found, it is detected [here](https://github.com/open-telemetry/opentelemetry-operator/blob/v0.90.0/cmd/otel-allocator/main.go#L175) and the [`SetCollectors()`](https://github.com/open-telemetry/opentelemetry-operator/blob/v0.90.0/cmd/otel-allocator/allocation/least_weighted.go#L272) function of the allocator is executed. It checks the differences of collectors between current and new states, finds a new one (`addition`) and therefore runs [`handleCollectors`](https://github.com/open-telemetry/opentelemetry-operator/blob/v0.90.0/cmd/otel-allocator/allocation/least_weighted.go#L179).

The critical part is [this](https://github.com/open-telemetry/opentelemetry-operator/blob/v0.90.0/cmd/otel-allocator/allocation/least_weighted.go#L188) `if case`. Since the number of the collectors is not zero, the flag `allocateTargets` remains `false` and therefore we never re-distribute the targets [here](https://github.com/open-telemetry/opentelemetry-operator/blob/v0.90.0/cmd/otel-allocator/allocation/least_weighted.go#L198). However, even we were to enter this loop, the rest of the implementation at the moment does not handle re-assigning targets from the old state of the collectors to the new one.

So as far as I've understood, if one wants to scale out, one needs to delete and re-create the allocator instance after introducing a new collector instance.

### Actual intention

The actual reason why I wanted to bring up this issue is that I am facing `OOMKilled` problems with the collectors due to uneven distribution of scrape targets and scaling them out seems not to help. Since the default value for the allocation strategy is defined as the `least-weighted` but not `consistent-hashing`, I assumed that the community encourages us to go for the `least-weighted` even though `consistent-hashing` strategy re-distributes all of the scrape targets in case a new collector instance is added.

I would be more than happy to get your experiences & recommendations.

### PR DESCRIPTION
**Pull request by utr1903**
**Description:**
As discussed in the linked issue:
- default value for allocation strategy is changed from `least-weighted` to `consistent-hashing`
- default value for filtering strategy is changed from `none` to `relabel-config`

**Link to tracking Issue:**
closes #2477

**Testing:**
None required.

**Documentation:**
None required.

### CODE REVIEW

**Review by jaronoff97 — COMMENTED** [review]
we'll also need to update the defaults [here](https://github.com/open-telemetry/opentelemetry-operator/blob/main/internal/manifests/targetallocator/configmap.go#L53-L61)

**Comment by swiatekm-sumo** [comment]
> we'll also need to update the defaults [here](https://github.com/open-telemetry/opentelemetry-operator/blob/main/internal/manifests/targetallocator/configmap.go#L53-L61)

While we're at it, let's set them in the CRDs as well. Like so:
```go
	// AllocationStrategy determines which strategy the target allocator should use for allocation.
	// The current options are least-weighted and consistent-hashing. The default option is least-weighted
	// +optional
	// +kubebuilder:default:="least-weighted"
	AllocationStrategy OpenTelemetryTargetAllocatorAllocationStrategy `json:"allocationStrategy,omitempty"`
```

**Comment by utr1903** [comment]
On it :)

**Comment by utr1903** [comment]
Would these changes be OK regarding the allocation strategy?

If yes, shall I do the exact same structure for the filtering strategy? We can:
- convert it from to string to enum
- add it's own file as in `allocation_strategy.go` -> `// +kubebuilder:validation:Enum="";relabel-config`
- set default value to `relabel-config` -> `// +kubebuilder:default:=relabel-config`

**Review by jaronoff97 — APPROVED** [review]

**Comment by jaronoff97** [comment]
@utr1903 Yes! that sounds great. I think i would maybe rename `allocation_strategy.go` to something like `allocator_constants.go` or something like that, otherwise sounds good :D

File: internal/manifests/targetallocator/configmap.go [hunk 1]
```diff
@@ -53,7 +53,7 @@ func ConfigMap(params manifests.Params) (*corev1.ConfigMap, error) {
 	if len(params.OtelCol.Spec.TargetAllocator.AllocationStrategy) > 0 {
 		taConfig["allocation_strategy"] = params.OtelCol.Spec.TargetAllocator.AllocationStrategy
 	} else {
-		taConfig["allocation_strategy"] = v1alpha1.OpenTelemetryTargetAllocatorAllocationStrategyLeastWeighted
+		taConfig["allocation_strategy"] = v1alpha1.OpenTelemetryTargetAllocatorAllocationStrategyConsistentHashing
 	}
```
Comment (swiatekm-sumo) [inline internal/manifests/targetallocator/configmap.go:57]: We've set the default on the CRD, so here we can unconditionally do:

```go
taConfig["allocation_strategy"] = params.OtelCol.Spec.TargetAllocator.AllocationStrategy
```

**Review by swiatekm-sumo — APPROVED** [review]
One minor comment, LGTM otherwise. :+1:

**Comment by jaronoff97** [comment]
@utr1903 looks like the hashes in the config are wrong, otherwise i think this is good to merge.

**Pull request merged**